File: PAT_Basic_Level_by_ruby/1017 A除以B/1017 A除以B.rb
# 1017 A除以B
#
# 题目描述：
#   本题要求计算 A/B，其中 A 是不超过 1000 位的正整数，B 是 1 位正整数。你需要输出商数 Q 和余数 R，使得 A = B \times Q + R 成立。
#
# 实现原理：
#   使用大整数除法，逐位进行计算；模拟手工除法的过程；处理商和余数
#

if __FILE__ == $PROGRAM_NAME
  a, b = gets.split
  b = b.to_i

  quotient = String.new
  remainder = 0
  first = true

  a.each_char do |ch|
    remainder = remainder * 10 + ch.to_i
    q = remainder / b
    unless first && q.zero?
      quotient << q.to_s
      first = false
    end
    remainder = remainder % b
  end

  quotient = '0' if quotient.empty?
  puts "#{quotient} #{remainder}"
end

File: PAT_Basic_Level_by_c/1017 A除以B/1017 A除以B.c
/*
 * 题目：1017 A除以B
 * 实现原理：
 *   计算高精度除法，A为不超过1000位的正整数，B为1位正整数，输出商和余数。
 *   算法步骤：
 *   1. 将A作为字符串读取，逐位进行除法运算
 *   2. 使用remainder记录当前余数，初始为0
 *   3. 对每一位数字：remainder = remainder * 10 + 当前位数字
 *   4. 计算当前位商：quotient = remainder / B
 *   5. 更新余数：remainder = remainder % B
 *   6. 使用first标志避免输出前导零
 *   特殊处理：若商为0（即A<B），需输出0。
 */
#include <stdio.h>
#include <string.h>

int main() {
    char a[1001];                   // A 最长 1000 位，用字符串存储
    int b;                          // 除数，为 1 位正整数
    scanf("%s %d", a, &b);
    int len = strlen(a);
    int first = 1;                  // 是否还未输出有效商位（用于去除前导零）
    int remainder = 0;              // 当前余数
    for (int i = 0; i < len; i++) { // 从高位到低位逐位做除法
        remainder = remainder * 10 + (a[i] - '0');   // 拼接当前位数字构成被除数
        int quotient = remainder / b;               // 当前位的商
        if (!first || quotient != 0) {              // 跳过商的前导零
            printf("%d", quotient);
            first = 0;
        }
        remainder %= b;                             // 更新余数
    }
    if (first) {                    // 商为 0（A < B 的情况），补输出 0
        printf("0");
    }
    printf(" %d\n", remainder);     // 输出余数
    return 0;
}